{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fa1abe55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pandas in /home/saad/.local/lib/python3.10/site-packages (1.5.3)\n",
      "Requirement already satisfied: numpy>=1.21.0 in /home/saad/.local/lib/python3.10/site-packages (from pandas) (1.24.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/saad/.local/lib/python3.10/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/saad/.local/lib/python3.10/site-packages (from pandas) (2022.7.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/saad/.local/lib/python3.10/site-packages (from python-dateutil>=2.8.1->pandas) (1.16.0)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: numpy in /home/saad/.local/lib/python3.10/site-packages (1.24.1)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pyarrow in /home/saad/.local/lib/python3.10/site-packages (11.0.0)\n",
      "Requirement already satisfied: numpy>=1.16.6 in /home/saad/.local/lib/python3.10/site-packages (from pyarrow) (1.24.1)\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import subprocess\n",
    "\n",
    "# implement pip as a subprocess:\n",
    "packages = ['pandas','numpy','pyarrow']\n",
    "for p in packages:\n",
    "    subprocess.check_call([sys.executable, '-m', 'pip', 'install',p])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "73931ac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import io\n",
    "import fnmatch\n",
    "import tarfile\n",
    "from ftplib import FTP,error_perm\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ab9fcf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "paths_to_df = [\n",
    " ('targets.feather','targets.tar.gz','pub/databases/opentargets/platform/21.11/output/etl/parquet/targets')\n",
    ",('diseases.feather','diseases.tar.gz','pub/databases/opentargets/platform/21.11/output/etl/parquet/diseases')\n",
    ",('evidence.feather','evidence.tar.gz','pub/databases/opentargets/platform/21.11/output/etl/parquet/evidence/sourceId=eva')\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "73c8a815",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_datasets_with_fallback():\n",
    "    try:\n",
    "        for feather_path,parquet_path,ftp_path in paths_to_df:\n",
    "            if os.path.isfile(feather_path):\n",
    "                yield pd.read_feather(feather_path)\n",
    "            elif os.path.isfile(parquet_path):\n",
    "                with tarfile.open(parquet_path, \"r:*\") as tar:\n",
    "                    df = pd.concat([pd.read_parquet(tar.extractfile(file)) for file in tar.getnames()])\n",
    "                    df.reset_index().to_feather(feather_path,compression = 'lz4')\n",
    "                    yield pd.read_feather(feather_path)\n",
    "            else:\n",
    "                ftp = FTP('ftp.ebi.ac.uk')\n",
    "                ftp.login()\n",
    "                ftp.cwd(path)\n",
    "                files = [filename for filename in ftp.nlst() if fnmatch.fnmatch(filename, '*.parquet')]\n",
    "                with tarfile.open(parquet_path, \"w:gz\") as fout:\n",
    "                    for file_ in files:\n",
    "                        buffer = io.BytesIO()\n",
    "                        ftp.retrbinary('RETR ' + str(file_), buffer.write)\n",
    "                        size = buffer.getbuffer().nbytes\n",
    "                        print(f\"Size of {file_} is: {size}\")\n",
    "                        tf = tarfile.TarInfo(file_)\n",
    "                        tf.size = size\n",
    "                        buffer.seek(0)\n",
    "                        fout.addfile(tf,buffer)\n",
    "                with tarfile.open(parquet_path, \"r:*\") as tar:\n",
    "                    df = pd.concat([pd.read_parquet(tar.extractfile(file)) for file in tar.getnames()])\n",
    "                    df.reset_index().to_feather(feather_path,compression = 'lz4')\n",
    "                    yield pd.read_feather(feather_path)\n",
    "    except error_perm as resp:\n",
    "        if str(resp) == \"550 No files found\":\n",
    "            print(\"No files in this directory\")\n",
    "        else:\n",
    "            raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7ff8ea5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>id</th>\n",
       "      <th>approvedSymbol</th>\n",
       "      <th>biotype</th>\n",
       "      <th>transcriptIds</th>\n",
       "      <th>genomicLocation</th>\n",
       "      <th>alternativeGenes</th>\n",
       "      <th>approvedName</th>\n",
       "      <th>go</th>\n",
       "      <th>hallmarks</th>\n",
       "      <th>...</th>\n",
       "      <th>obsoleteNames</th>\n",
       "      <th>constraint</th>\n",
       "      <th>tep</th>\n",
       "      <th>proteinIds</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chemicalProbes</th>\n",
       "      <th>homologues</th>\n",
       "      <th>tractability</th>\n",
       "      <th>safetyLiabilities</th>\n",
       "      <th>pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000020219</td>\n",
       "      <td>CCT8L1P</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>[ENST00000465400]</td>\n",
       "      <td>{'chromosome': '7', 'end': 152447150, 'start':...</td>\n",
       "      <td>None</td>\n",
       "      <td>chaperonin containing TCP1 subunit 8 like 1, p...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'label': 'chaperonin containing TCP1, subuni...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'A6NM43', 'source': 'uniprot_obsolete'}]</td>\n",
       "      <td>[{'id': '32153', 'source': 'HGNC'}, {'id': 'IP...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000059588</td>\n",
       "      <td>TARBP1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000496673, ENST00000484454, ENST0000048...</td>\n",
       "      <td>{'chromosome': '1', 'end': 234479179, 'start':...</td>\n",
       "      <td>None</td>\n",
       "      <td>TAR (HIV-1) RNA binding protein 1</td>\n",
       "      <td>[{'aspect': 'F', 'ecoId': 'ECO:0007669', 'evid...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'label': 'Tar (HIV-1) RNA binding protein 1'...</td>\n",
       "      <td>[{'constraintType': 'syn', 'exp': 272.51000976...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'Q13395', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '11568', 'source': 'HGNC'}, {'id': '2H...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'homologyType': 'other_paralog', 'isHighConf...</td>\n",
       "      <td>[{'id': 'Approved Drug', 'modality': 'SM', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000070182</td>\n",
       "      <td>SPTB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000553938, ENST00000389720, ENST0000055...</td>\n",
       "      <td>{'chromosome': '14', 'end': 64879907, 'start':...</td>\n",
       "      <td>None</td>\n",
       "      <td>spectrin beta, erythrocytic</td>\n",
       "      <td>[{'aspect': 'P', 'ecoId': 'ECO:0007669', 'evid...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'constraintType': 'syn', 'exp': 586.70001220...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'P11277', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '11274', 'source': 'HGNC'}, {'id': '1S...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'homologyType': 'other_paralog', 'isHighConf...</td>\n",
       "      <td>[{'id': 'Approved Drug', 'modality': 'SM', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'pathway': 'COPI-mediated anterograde transp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000070366</td>\n",
       "      <td>SMG6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000354901, ENST00000570756, ENST0000057...</td>\n",
       "      <td>{'chromosome': '17', 'end': 2303785, 'start': ...</td>\n",
       "      <td>None</td>\n",
       "      <td>SMG6 nonsense mediated mRNA decay factor</td>\n",
       "      <td>[{'aspect': 'P', 'ecoId': 'ECO:0000314', 'evid...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'label': 'chromosome 17 open reading frame 3...</td>\n",
       "      <td>[{'constraintType': 'syn', 'exp': 321.14001464...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'Q86US8', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '17809', 'source': 'HGNC'}, {'id': '2D...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'homologyType': 'ortholog_one2one', 'isHighC...</td>\n",
       "      <td>[{'id': 'Approved Drug', 'modality': 'SM', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'pathway': 'Nonsense Mediated Decay (NMD) en...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000072071</td>\n",
       "      <td>ADGRL1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000361434, ENST00000589616, ENST0000059...</td>\n",
       "      <td>{'chromosome': '19', 'end': 14206187, 'start':...</td>\n",
       "      <td>[ENSG00000288324]</td>\n",
       "      <td>adhesion G protein-coupled receptor L1</td>\n",
       "      <td>[{'aspect': 'F', 'ecoId': 'ECO:0000250', 'evid...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'label': 'latrophilin 1', 'source': 'HGNC'}]</td>\n",
       "      <td>[{'constraintType': 'syn', 'exp': 424.92999267...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'O94910', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '20973', 'source': 'HGNC'}, {'id': 'IP...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'homologyType': 'other_paralog', 'isHighConf...</td>\n",
       "      <td>[{'id': 'Approved Drug', 'modality': 'SM', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60631</th>\n",
       "      <td>282</td>\n",
       "      <td>ENSG00000287783</td>\n",
       "      <td>ENSG00000287783</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>[ENST00000667965, ENST00000662579]</td>\n",
       "      <td>{'chromosome': '5', 'end': 87774070, 'start': ...</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60632</th>\n",
       "      <td>283</td>\n",
       "      <td>ENSG00000287887</td>\n",
       "      <td>ENSG00000287887</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>[ENST00000662637]</td>\n",
       "      <td>{'chromosome': '13', 'end': 24755201, 'start':...</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60633</th>\n",
       "      <td>284</td>\n",
       "      <td>ENSG00000287908</td>\n",
       "      <td>ENSG00000287908</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000477314, ENST00000474359]</td>\n",
       "      <td>{'chromosome': '12', 'end': 57626125, 'start':...</td>\n",
       "      <td>None</td>\n",
       "      <td>novel protein</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'E9PIH7', 'source': 'uniprot_trembl'}]</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'homologyType': 'within_species_paralog', 'i...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60634</th>\n",
       "      <td>285</td>\n",
       "      <td>ENSG00000287915</td>\n",
       "      <td>ENSG00000287915</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>[ENST00000662300]</td>\n",
       "      <td>{'chromosome': '15', 'end': 60223272, 'start':...</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60635</th>\n",
       "      <td>286</td>\n",
       "      <td>ENSG00000288026</td>\n",
       "      <td>ENSG00000288026</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>[ENST00000667168]</td>\n",
       "      <td>{'chromosome': '16', 'end': 48055099, 'start':...</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60636 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index               id   approvedSymbol               biotype  \\\n",
       "0          0  ENSG00000020219          CCT8L1P  processed_pseudogene   \n",
       "1          1  ENSG00000059588           TARBP1        protein_coding   \n",
       "2          2  ENSG00000070182             SPTB        protein_coding   \n",
       "3          3  ENSG00000070366             SMG6        protein_coding   \n",
       "4          4  ENSG00000072071           ADGRL1        protein_coding   \n",
       "...      ...              ...              ...                   ...   \n",
       "60631    282  ENSG00000287783  ENSG00000287783                lncRNA   \n",
       "60632    283  ENSG00000287887  ENSG00000287887                lncRNA   \n",
       "60633    284  ENSG00000287908  ENSG00000287908        protein_coding   \n",
       "60634    285  ENSG00000287915  ENSG00000287915                lncRNA   \n",
       "60635    286  ENSG00000288026  ENSG00000288026                lncRNA   \n",
       "\n",
       "                                           transcriptIds  \\\n",
       "0                                      [ENST00000465400]   \n",
       "1      [ENST00000496673, ENST00000484454, ENST0000048...   \n",
       "2      [ENST00000553938, ENST00000389720, ENST0000055...   \n",
       "3      [ENST00000354901, ENST00000570756, ENST0000057...   \n",
       "4      [ENST00000361434, ENST00000589616, ENST0000059...   \n",
       "...                                                  ...   \n",
       "60631                 [ENST00000667965, ENST00000662579]   \n",
       "60632                                  [ENST00000662637]   \n",
       "60633                 [ENST00000477314, ENST00000474359]   \n",
       "60634                                  [ENST00000662300]   \n",
       "60635                                  [ENST00000667168]   \n",
       "\n",
       "                                         genomicLocation   alternativeGenes  \\\n",
       "0      {'chromosome': '7', 'end': 152447150, 'start':...               None   \n",
       "1      {'chromosome': '1', 'end': 234479179, 'start':...               None   \n",
       "2      {'chromosome': '14', 'end': 64879907, 'start':...               None   \n",
       "3      {'chromosome': '17', 'end': 2303785, 'start': ...               None   \n",
       "4      {'chromosome': '19', 'end': 14206187, 'start':...  [ENSG00000288324]   \n",
       "...                                                  ...                ...   \n",
       "60631  {'chromosome': '5', 'end': 87774070, 'start': ...               None   \n",
       "60632  {'chromosome': '13', 'end': 24755201, 'start':...               None   \n",
       "60633  {'chromosome': '12', 'end': 57626125, 'start':...               None   \n",
       "60634  {'chromosome': '15', 'end': 60223272, 'start':...               None   \n",
       "60635  {'chromosome': '16', 'end': 48055099, 'start':...               None   \n",
       "\n",
       "                                            approvedName  \\\n",
       "0      chaperonin containing TCP1 subunit 8 like 1, p...   \n",
       "1                      TAR (HIV-1) RNA binding protein 1   \n",
       "2                            spectrin beta, erythrocytic   \n",
       "3               SMG6 nonsense mediated mRNA decay factor   \n",
       "4                 adhesion G protein-coupled receptor L1   \n",
       "...                                                  ...   \n",
       "60631                                   novel transcript   \n",
       "60632                                   novel transcript   \n",
       "60633                                      novel protein   \n",
       "60634                                   novel transcript   \n",
       "60635                                   novel transcript   \n",
       "\n",
       "                                                      go hallmarks  ...  \\\n",
       "0                                                   None      None  ...   \n",
       "1      [{'aspect': 'F', 'ecoId': 'ECO:0007669', 'evid...      None  ...   \n",
       "2      [{'aspect': 'P', 'ecoId': 'ECO:0007669', 'evid...      None  ...   \n",
       "3      [{'aspect': 'P', 'ecoId': 'ECO:0000314', 'evid...      None  ...   \n",
       "4      [{'aspect': 'F', 'ecoId': 'ECO:0000250', 'evid...      None  ...   \n",
       "...                                                  ...       ...  ...   \n",
       "60631                                               None      None  ...   \n",
       "60632                                               None      None  ...   \n",
       "60633                                               None      None  ...   \n",
       "60634                                               None      None  ...   \n",
       "60635                                               None      None  ...   \n",
       "\n",
       "                                           obsoleteNames  \\\n",
       "0      [{'label': 'chaperonin containing TCP1, subuni...   \n",
       "1      [{'label': 'Tar (HIV-1) RNA binding protein 1'...   \n",
       "2                                                     []   \n",
       "3      [{'label': 'chromosome 17 open reading frame 3...   \n",
       "4         [{'label': 'latrophilin 1', 'source': 'HGNC'}]   \n",
       "...                                                  ...   \n",
       "60631                                                 []   \n",
       "60632                                                 []   \n",
       "60633                                                 []   \n",
       "60634                                                 []   \n",
       "60635                                                 []   \n",
       "\n",
       "                                              constraint   tep  \\\n",
       "0                                                   None  None   \n",
       "1      [{'constraintType': 'syn', 'exp': 272.51000976...  None   \n",
       "2      [{'constraintType': 'syn', 'exp': 586.70001220...  None   \n",
       "3      [{'constraintType': 'syn', 'exp': 321.14001464...  None   \n",
       "4      [{'constraintType': 'syn', 'exp': 424.92999267...  None   \n",
       "...                                                  ...   ...   \n",
       "60631                                               None  None   \n",
       "60632                                               None  None   \n",
       "60633                                               None  None   \n",
       "60634                                               None  None   \n",
       "60635                                               None  None   \n",
       "\n",
       "                                              proteinIds  \\\n",
       "0       [{'id': 'A6NM43', 'source': 'uniprot_obsolete'}]   \n",
       "1      [{'id': 'Q13395', 'source': 'uniprot_swissprot...   \n",
       "2      [{'id': 'P11277', 'source': 'uniprot_swissprot...   \n",
       "3      [{'id': 'Q86US8', 'source': 'uniprot_swissprot...   \n",
       "4      [{'id': 'O94910', 'source': 'uniprot_swissprot...   \n",
       "...                                                  ...   \n",
       "60631                                               None   \n",
       "60632                                               None   \n",
       "60633     [{'id': 'E9PIH7', 'source': 'uniprot_trembl'}]   \n",
       "60634                                               None   \n",
       "60635                                               None   \n",
       "\n",
       "                                                 dbXrefs chemicalProbes  \\\n",
       "0      [{'id': '32153', 'source': 'HGNC'}, {'id': 'IP...           None   \n",
       "1      [{'id': '11568', 'source': 'HGNC'}, {'id': '2H...           None   \n",
       "2      [{'id': '11274', 'source': 'HGNC'}, {'id': '1S...           None   \n",
       "3      [{'id': '17809', 'source': 'HGNC'}, {'id': '2D...           None   \n",
       "4      [{'id': '20973', 'source': 'HGNC'}, {'id': 'IP...           None   \n",
       "...                                                  ...            ...   \n",
       "60631                                                 []           None   \n",
       "60632                                                 []           None   \n",
       "60633                                                 []           None   \n",
       "60634                                                 []           None   \n",
       "60635                                                 []           None   \n",
       "\n",
       "                                              homologues  \\\n",
       "0                                                   None   \n",
       "1      [{'homologyType': 'other_paralog', 'isHighConf...   \n",
       "2      [{'homologyType': 'other_paralog', 'isHighConf...   \n",
       "3      [{'homologyType': 'ortholog_one2one', 'isHighC...   \n",
       "4      [{'homologyType': 'other_paralog', 'isHighConf...   \n",
       "...                                                  ...   \n",
       "60631                                               None   \n",
       "60632                                               None   \n",
       "60633  [{'homologyType': 'within_species_paralog', 'i...   \n",
       "60634                                               None   \n",
       "60635                                               None   \n",
       "\n",
       "                                            tractability safetyLiabilities  \\\n",
       "0                                                   None              None   \n",
       "1      [{'id': 'Approved Drug', 'modality': 'SM', 'va...              None   \n",
       "2      [{'id': 'Approved Drug', 'modality': 'SM', 'va...              None   \n",
       "3      [{'id': 'Approved Drug', 'modality': 'SM', 'va...              None   \n",
       "4      [{'id': 'Approved Drug', 'modality': 'SM', 'va...              None   \n",
       "...                                                  ...               ...   \n",
       "60631                                               None              None   \n",
       "60632                                               None              None   \n",
       "60633                                               None              None   \n",
       "60634                                               None              None   \n",
       "60635                                               None              None   \n",
       "\n",
       "                                                pathways  \n",
       "0                                                   None  \n",
       "1                                                   None  \n",
       "2      [{'pathway': 'COPI-mediated anterograde transp...  \n",
       "3      [{'pathway': 'Nonsense Mediated Decay (NMD) en...  \n",
       "4                                                   None  \n",
       "...                                                  ...  \n",
       "60631                                               None  \n",
       "60632                                               None  \n",
       "60633                                               None  \n",
       "60634                                               None  \n",
       "60635                                               None  \n",
       "\n",
       "[60636 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>id</th>\n",
       "      <th>code</th>\n",
       "      <th>dbXRefs</th>\n",
       "      <th>description</th>\n",
       "      <th>name</th>\n",
       "      <th>directLocationIds</th>\n",
       "      <th>obsoleteTerms</th>\n",
       "      <th>parents</th>\n",
       "      <th>sko</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>ancestors</th>\n",
       "      <th>descendants</th>\n",
       "      <th>children</th>\n",
       "      <th>therapeuticAreas</th>\n",
       "      <th>indirectLocationIds</th>\n",
       "      <th>ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>DOID_7551</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_7551</td>\n",
       "      <td>[COHD:433417, ICD9:098.89, MeSH:D006069, NCIt:...</td>\n",
       "      <td>A primary bacterial infectious disease that is...</td>\n",
       "      <td>gonorrhea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0003955, MONDO_0000314]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[EFO_0000512, EFO_0009663, MONDO_0100336, OTAR...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[MONDO_0100336, OTAR_0000017, EFO_0005741, EFO...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': True, 'so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HP_0000031</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HP_0000031</td>\n",
       "      <td>[UMLS:C0014534, SNOMEDCT_US:31070006, MSH:D004...</td>\n",
       "      <td>The presence of inflammation of the epididymis.</td>\n",
       "      <td>Epididymitis</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[HP_0012649, HP_0000022]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[HP_0000022, HP_0000078, EFO_0000651, HP_00126...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[EFO_0000651]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': True, 'so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>HP_0000857</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HP_0000857</td>\n",
       "      <td>[UMLS:C3278636]</td>\n",
       "      <td>None</td>\n",
       "      <td>Neonatal insulin-dependent diabetes mellitus</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0000400, HP_0001952, HP_0000818]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[HP_0000818, HP_0000118, OTAR_0000020, HP_0011...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[OTAR_0000020, EFO_0000651, EFO_0001379, EFO_0...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': True, 'so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>HP_0001693</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HP_0001693</td>\n",
       "      <td>[UMLS:C0232180, SNOMEDCT_US:442119001]</td>\n",
       "      <td>Pattern of blood flow in the heart that deviat...</td>\n",
       "      <td>Cardiac shunt</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[HP_0011028]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0000651, HP_0011028, HP_0000118, HP_00110...</td>\n",
       "      <td>[HP_0012382]</td>\n",
       "      <td>[HP_0012382]</td>\n",
       "      <td>[EFO_0000651]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>HP_0001730</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HP_0001730</td>\n",
       "      <td>[UMLS:C1842138]</td>\n",
       "      <td>A progressive form of hearing impairment.</td>\n",
       "      <td>Progressive hearing impairment</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[HP_0000364]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[HP_0000364, EFO_0000651, HP_0000598, HP_0000118]</td>\n",
       "      <td>[HP_0000408]</td>\n",
       "      <td>[HP_0000408]</td>\n",
       "      <td>[EFO_0000651]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18701</th>\n",
       "      <td>243</td>\n",
       "      <td>Orphanet_93322</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_93322</td>\n",
       "      <td>[MESH:C535563, ICD10:Q72.5, MONDO:0010144, OMI...</td>\n",
       "      <td>Tibial hemimelia is a rare congenital anomaly ...</td>\n",
       "      <td>Tibial hemimelia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[Orphanet_2130]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[EFO_0004260, Orphanet_404571, OTAR_0000006, O...</td>\n",
       "      <td>[Orphanet_295079, Orphanet_295077]</td>\n",
       "      <td>[Orphanet_295077, Orphanet_295079]</td>\n",
       "      <td>[OTAR_0000006, OTAR_0000018]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18702</th>\n",
       "      <td>244</td>\n",
       "      <td>Orphanet_93925</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_93925</td>\n",
       "      <td>[UMLS:C0431363, ICD10:Q04.2, MONDO:0019757, SC...</td>\n",
       "      <td>Alobar holoprosencephaly is the most severe cl...</td>\n",
       "      <td>Alobar holoprosencephaly</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[Orphanet_2162]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[Orphanet_183763, MONDO_0002025, MONDO_0019042...</td>\n",
       "      <td>[MONDO_0012322]</td>\n",
       "      <td>[MONDO_0012322]</td>\n",
       "      <td>[MONDO_0002025, EFO_0000618, OTAR_0000018, EFO...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18703</th>\n",
       "      <td>245</td>\n",
       "      <td>Orphanet_98539</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_98539</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>Early-onset ataxia with dementia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[Orphanet_98538]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[MONDO_0002025, Orphanet_183500, OTAR_0000018,...</td>\n",
       "      <td>[Orphanet_247815, EFO_0009016, Orphanet_284282...</td>\n",
       "      <td>[Orphanet_1172]</td>\n",
       "      <td>[MONDO_0002025, OTAR_0000018, EFO_0000618]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18704</th>\n",
       "      <td>246</td>\n",
       "      <td>Orphanet_98636</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_98636</td>\n",
       "      <td>[MONDO:0020220]</td>\n",
       "      <td>None</td>\n",
       "      <td>Corneoiridogoniodysgenesis</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[Orphanet_98632]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[OTAR_0000018, Orphanet_183557, Orphanet_98632...</td>\n",
       "      <td>[Orphanet_98941, Orphanet_708, Orphanet_101033]</td>\n",
       "      <td>[Orphanet_708]</td>\n",
       "      <td>[OTAR_0000018, MONDO_0024458]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18705</th>\n",
       "      <td>247</td>\n",
       "      <td>Orphanet_99951</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_99951</td>\n",
       "      <td>[OMIM:605253, DOID:0110195, GARD:0006170, MeSH...</td>\n",
       "      <td>Charcot-Marie-Tooth disease type 4E (CMT4E) is...</td>\n",
       "      <td>Charcot-Marie-Tooth disease type 4E</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[MONDO_0033352, Orphanet_64749]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[Orphanet_98497, Orphanet_183500, OTAR_0000006...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[OTAR_0000006, EFO_0000618, OTAR_0000018]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': True, 'so...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18706 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index              id                                       code  \\\n",
       "0          0       DOID_7551   http://purl.obolibrary.org/obo/DOID_7551   \n",
       "1          1      HP_0000031  http://purl.obolibrary.org/obo/HP_0000031   \n",
       "2          2      HP_0000857  http://purl.obolibrary.org/obo/HP_0000857   \n",
       "3          3      HP_0001693  http://purl.obolibrary.org/obo/HP_0001693   \n",
       "4          4      HP_0001730  http://purl.obolibrary.org/obo/HP_0001730   \n",
       "...      ...             ...                                        ...   \n",
       "18701    243  Orphanet_93322   http://www.orpha.net/ORDO/Orphanet_93322   \n",
       "18702    244  Orphanet_93925   http://www.orpha.net/ORDO/Orphanet_93925   \n",
       "18703    245  Orphanet_98539   http://www.orpha.net/ORDO/Orphanet_98539   \n",
       "18704    246  Orphanet_98636   http://www.orpha.net/ORDO/Orphanet_98636   \n",
       "18705    247  Orphanet_99951   http://www.orpha.net/ORDO/Orphanet_99951   \n",
       "\n",
       "                                                 dbXRefs  \\\n",
       "0      [COHD:433417, ICD9:098.89, MeSH:D006069, NCIt:...   \n",
       "1      [UMLS:C0014534, SNOMEDCT_US:31070006, MSH:D004...   \n",
       "2                                        [UMLS:C3278636]   \n",
       "3                 [UMLS:C0232180, SNOMEDCT_US:442119001]   \n",
       "4                                        [UMLS:C1842138]   \n",
       "...                                                  ...   \n",
       "18701  [MESH:C535563, ICD10:Q72.5, MONDO:0010144, OMI...   \n",
       "18702  [UMLS:C0431363, ICD10:Q04.2, MONDO:0019757, SC...   \n",
       "18703                                                 []   \n",
       "18704                                    [MONDO:0020220]   \n",
       "18705  [OMIM:605253, DOID:0110195, GARD:0006170, MeSH...   \n",
       "\n",
       "                                             description  \\\n",
       "0      A primary bacterial infectious disease that is...   \n",
       "1        The presence of inflammation of the epididymis.   \n",
       "2                                                   None   \n",
       "3      Pattern of blood flow in the heart that deviat...   \n",
       "4              A progressive form of hearing impairment.   \n",
       "...                                                  ...   \n",
       "18701  Tibial hemimelia is a rare congenital anomaly ...   \n",
       "18702  Alobar holoprosencephaly is the most severe cl...   \n",
       "18703                                               None   \n",
       "18704                                               None   \n",
       "18705  Charcot-Marie-Tooth disease type 4E (CMT4E) is...   \n",
       "\n",
       "                                               name directLocationIds  \\\n",
       "0                                         gonorrhea              None   \n",
       "1                                      Epididymitis              None   \n",
       "2      Neonatal insulin-dependent diabetes mellitus              None   \n",
       "3                                     Cardiac shunt              None   \n",
       "4                    Progressive hearing impairment              None   \n",
       "...                                             ...               ...   \n",
       "18701                              Tibial hemimelia              None   \n",
       "18702                      Alobar holoprosencephaly              None   \n",
       "18703              Early-onset ataxia with dementia              None   \n",
       "18704                    Corneoiridogoniodysgenesis              None   \n",
       "18705           Charcot-Marie-Tooth disease type 4E              None   \n",
       "\n",
       "      obsoleteTerms                                parents   sko  \\\n",
       "0              None           [EFO_0003955, MONDO_0000314]  None   \n",
       "1              None               [HP_0012649, HP_0000022]  None   \n",
       "2              None  [EFO_0000400, HP_0001952, HP_0000818]  None   \n",
       "3              None                           [HP_0011028]  None   \n",
       "4              None                           [HP_0000364]  None   \n",
       "...             ...                                    ...   ...   \n",
       "18701          None                        [Orphanet_2130]  None   \n",
       "18702          None                        [Orphanet_2162]  None   \n",
       "18703          None                       [Orphanet_98538]  None   \n",
       "18704          None                       [Orphanet_98632]  None   \n",
       "18705          None        [MONDO_0033352, Orphanet_64749]  None   \n",
       "\n",
       "                                                synonyms  \\\n",
       "0      {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "1                                                   None   \n",
       "2                                                   None   \n",
       "3                                                   None   \n",
       "4      {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "...                                                  ...   \n",
       "18701  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "18702  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "18703                                               None   \n",
       "18704  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "18705  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "\n",
       "                                               ancestors  \\\n",
       "0      [EFO_0000512, EFO_0009663, MONDO_0100336, OTAR...   \n",
       "1      [HP_0000022, HP_0000078, EFO_0000651, HP_00126...   \n",
       "2      [HP_0000818, HP_0000118, OTAR_0000020, HP_0011...   \n",
       "3      [EFO_0000651, HP_0011028, HP_0000118, HP_00110...   \n",
       "4      [HP_0000364, EFO_0000651, HP_0000598, HP_0000118]   \n",
       "...                                                  ...   \n",
       "18701  [EFO_0004260, Orphanet_404571, OTAR_0000006, O...   \n",
       "18702  [Orphanet_183763, MONDO_0002025, MONDO_0019042...   \n",
       "18703  [MONDO_0002025, Orphanet_183500, OTAR_0000018,...   \n",
       "18704  [OTAR_0000018, Orphanet_183557, Orphanet_98632...   \n",
       "18705  [Orphanet_98497, Orphanet_183500, OTAR_0000006...   \n",
       "\n",
       "                                             descendants  \\\n",
       "0                                                     []   \n",
       "1                                                     []   \n",
       "2                                                     []   \n",
       "3                                           [HP_0012382]   \n",
       "4                                           [HP_0000408]   \n",
       "...                                                  ...   \n",
       "18701                 [Orphanet_295079, Orphanet_295077]   \n",
       "18702                                    [MONDO_0012322]   \n",
       "18703  [Orphanet_247815, EFO_0009016, Orphanet_284282...   \n",
       "18704    [Orphanet_98941, Orphanet_708, Orphanet_101033]   \n",
       "18705                                                 []   \n",
       "\n",
       "                                 children  \\\n",
       "0                                      []   \n",
       "1                                      []   \n",
       "2                                      []   \n",
       "3                            [HP_0012382]   \n",
       "4                            [HP_0000408]   \n",
       "...                                   ...   \n",
       "18701  [Orphanet_295077, Orphanet_295079]   \n",
       "18702                     [MONDO_0012322]   \n",
       "18703                     [Orphanet_1172]   \n",
       "18704                      [Orphanet_708]   \n",
       "18705                                  []   \n",
       "\n",
       "                                        therapeuticAreas indirectLocationIds  \\\n",
       "0      [MONDO_0100336, OTAR_0000017, EFO_0005741, EFO...                None   \n",
       "1                                          [EFO_0000651]                None   \n",
       "2      [OTAR_0000020, EFO_0000651, EFO_0001379, EFO_0...                None   \n",
       "3                                          [EFO_0000651]                None   \n",
       "4                                          [EFO_0000651]                None   \n",
       "...                                                  ...                 ...   \n",
       "18701                       [OTAR_0000006, OTAR_0000018]                None   \n",
       "18702  [MONDO_0002025, EFO_0000618, OTAR_0000018, EFO...                None   \n",
       "18703         [MONDO_0002025, OTAR_0000018, EFO_0000618]                None   \n",
       "18704                      [OTAR_0000018, MONDO_0024458]                None   \n",
       "18705          [OTAR_0000006, EFO_0000618, OTAR_0000018]                None   \n",
       "\n",
       "                                                ontology  \n",
       "0      {'isTherapeuticArea': False, 'leaf': True, 'so...  \n",
       "1      {'isTherapeuticArea': False, 'leaf': True, 'so...  \n",
       "2      {'isTherapeuticArea': False, 'leaf': True, 'so...  \n",
       "3      {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "4      {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "...                                                  ...  \n",
       "18701  {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "18702  {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "18703  {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "18704  {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "18705  {'isTherapeuticArea': False, 'leaf': True, 'so...  \n",
       "\n",
       "[18706 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>datasourceId</th>\n",
       "      <th>targetId</th>\n",
       "      <th>alleleOrigins</th>\n",
       "      <th>allelicRequirements</th>\n",
       "      <th>beta</th>\n",
       "      <th>betaConfidenceIntervalLower</th>\n",
       "      <th>betaConfidenceIntervalUpper</th>\n",
       "      <th>biologicalModelAllelicComposition</th>\n",
       "      <th>biologicalModelGeneticBackground</th>\n",
       "      <th>...</th>\n",
       "      <th>targetModulation</th>\n",
       "      <th>textMiningSentences</th>\n",
       "      <th>urls</th>\n",
       "      <th>variantAminoacidDescriptions</th>\n",
       "      <th>variantFunctionalConsequenceId</th>\n",
       "      <th>variantId</th>\n",
       "      <th>variantRsId</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000168646</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001819</td>\n",
       "      <td>17_65537821_C_T</td>\n",
       "      <td>rs1053887162</td>\n",
       "      <td>Orphanet_300576</td>\n",
       "      <td>000b404a55aab3d6e7df8d43501ec930a3f92fd8</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000183873</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001583</td>\n",
       "      <td>3_38550706_T_C</td>\n",
       "      <td>rs1038605800</td>\n",
       "      <td>Orphanet_130</td>\n",
       "      <td>0010135b87920c6d0c95690fcc5380217479fe6b</td>\n",
       "      <td>0.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000101605</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001589</td>\n",
       "      <td>18_3129232_CTCTG_C</td>\n",
       "      <td>None</td>\n",
       "      <td>EFO_0000538</td>\n",
       "      <td>002d610963d28704f2b207b4b14c47f2b55df14b</td>\n",
       "      <td>0.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000169247</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001624</td>\n",
       "      <td>5_148998409_A_T</td>\n",
       "      <td>rs545169695</td>\n",
       "      <td>EFO_0004143</td>\n",
       "      <td>0030b182518da8abfc29d349fe478bfb66903132</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000096696</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001574</td>\n",
       "      <td>6_7559224_A_G</td>\n",
       "      <td>None</td>\n",
       "      <td>Orphanet_247</td>\n",
       "      <td>003bb0a89a48ab62d4f57b2c0b6f1a29f099b4a3</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210868</th>\n",
       "      <td>5543</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000168646</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001819</td>\n",
       "      <td>17_65558462_G_A</td>\n",
       "      <td>None</td>\n",
       "      <td>Orphanet_300576</td>\n",
       "      <td>ffdfa4b3c9a9212774a86a99f39beed20318d1a6</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210869</th>\n",
       "      <td>5544</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000169562</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001589</td>\n",
       "      <td>X_71224504_GC_G</td>\n",
       "      <td>rs1602350003</td>\n",
       "      <td>Orphanet_166</td>\n",
       "      <td>ffe4c57006027b9016c07f390c505e9bf8c5ab5a</td>\n",
       "      <td>0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210870</th>\n",
       "      <td>5545</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000062822</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001819</td>\n",
       "      <td>19_50402359_C_T</td>\n",
       "      <td>rs746475325</td>\n",
       "      <td>Orphanet_140162</td>\n",
       "      <td>ffe88b1f2bbe8aaebff80111bde247fefbf090a3</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210871</th>\n",
       "      <td>5546</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000155657</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001819</td>\n",
       "      <td>2_178729466_G_A</td>\n",
       "      <td>None</td>\n",
       "      <td>Orphanet_263</td>\n",
       "      <td>ffeaf543ce77a973514d395d2df49e94de8ec95e</td>\n",
       "      <td>0.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210872</th>\n",
       "      <td>5547</td>\n",
       "      <td>eva</td>\n",
       "      <td>ENSG00000204264</td>\n",
       "      <td>[germline]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SO_0001624</td>\n",
       "      <td>6_32840843_G_A</td>\n",
       "      <td>None</td>\n",
       "      <td>Orphanet_324999</td>\n",
       "      <td>fff395754e40f4f71bcd0209715dcca7e6ef7bad</td>\n",
       "      <td>0.32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210873 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index datasourceId         targetId alleleOrigins allelicRequirements  \\\n",
       "0           0          eva  ENSG00000168646    [germline]                None   \n",
       "1           1          eva  ENSG00000183873    [germline]                None   \n",
       "2           2          eva  ENSG00000101605    [germline]                None   \n",
       "3           3          eva  ENSG00000169247    [germline]                None   \n",
       "4           4          eva  ENSG00000096696    [germline]                None   \n",
       "...       ...          ...              ...           ...                 ...   \n",
       "210868   5543          eva  ENSG00000168646    [germline]                None   \n",
       "210869   5544          eva  ENSG00000169562    [germline]                None   \n",
       "210870   5545          eva  ENSG00000062822    [germline]                None   \n",
       "210871   5546          eva  ENSG00000155657    [germline]                None   \n",
       "210872   5547          eva  ENSG00000204264    [germline]                None   \n",
       "\n",
       "        beta  betaConfidenceIntervalLower  betaConfidenceIntervalUpper  \\\n",
       "0        NaN                          NaN                          NaN   \n",
       "1        NaN                          NaN                          NaN   \n",
       "2        NaN                          NaN                          NaN   \n",
       "3        NaN                          NaN                          NaN   \n",
       "4        NaN                          NaN                          NaN   \n",
       "...      ...                          ...                          ...   \n",
       "210868   NaN                          NaN                          NaN   \n",
       "210869   NaN                          NaN                          NaN   \n",
       "210870   NaN                          NaN                          NaN   \n",
       "210871   NaN                          NaN                          NaN   \n",
       "210872   NaN                          NaN                          NaN   \n",
       "\n",
       "       biologicalModelAllelicComposition biologicalModelGeneticBackground  \\\n",
       "0                                   None                             None   \n",
       "1                                   None                             None   \n",
       "2                                   None                             None   \n",
       "3                                   None                             None   \n",
       "4                                   None                             None   \n",
       "...                                  ...                              ...   \n",
       "210868                              None                             None   \n",
       "210869                              None                             None   \n",
       "210870                              None                             None   \n",
       "210871                              None                             None   \n",
       "210872                              None                             None   \n",
       "\n",
       "        ... targetModulation textMiningSentences  urls  \\\n",
       "0       ...             None                None  None   \n",
       "1       ...             None                None  None   \n",
       "2       ...             None                None  None   \n",
       "3       ...             None                None  None   \n",
       "4       ...             None                None  None   \n",
       "...     ...              ...                 ...   ...   \n",
       "210868  ...             None                None  None   \n",
       "210869  ...             None                None  None   \n",
       "210870  ...             None                None  None   \n",
       "210871  ...             None                None  None   \n",
       "210872  ...             None                None  None   \n",
       "\n",
       "       variantAminoacidDescriptions  variantFunctionalConsequenceId  \\\n",
       "0                              None                      SO_0001819   \n",
       "1                              None                      SO_0001583   \n",
       "2                              None                      SO_0001589   \n",
       "3                              None                      SO_0001624   \n",
       "4                              None                      SO_0001574   \n",
       "...                             ...                             ...   \n",
       "210868                         None                      SO_0001819   \n",
       "210869                         None                      SO_0001589   \n",
       "210870                         None                      SO_0001819   \n",
       "210871                         None                      SO_0001819   \n",
       "210872                         None                      SO_0001624   \n",
       "\n",
       "                 variantId   variantRsId        diseaseId  \\\n",
       "0          17_65537821_C_T  rs1053887162  Orphanet_300576   \n",
       "1           3_38550706_T_C  rs1038605800     Orphanet_130   \n",
       "2       18_3129232_CTCTG_C          None      EFO_0000538   \n",
       "3          5_148998409_A_T   rs545169695      EFO_0004143   \n",
       "4            6_7559224_A_G          None     Orphanet_247   \n",
       "...                    ...           ...              ...   \n",
       "210868     17_65558462_G_A          None  Orphanet_300576   \n",
       "210869     X_71224504_GC_G  rs1602350003     Orphanet_166   \n",
       "210870     19_50402359_C_T   rs746475325  Orphanet_140162   \n",
       "210871     2_178729466_G_A          None     Orphanet_263   \n",
       "210872      6_32840843_G_A          None  Orphanet_324999   \n",
       "\n",
       "                                              id score  \n",
       "0       000b404a55aab3d6e7df8d43501ec930a3f92fd8  0.02  \n",
       "1       0010135b87920c6d0c95690fcc5380217479fe6b  0.32  \n",
       "2       002d610963d28704f2b207b4b14c47f2b55df14b  0.32  \n",
       "3       0030b182518da8abfc29d349fe478bfb66903132  0.02  \n",
       "4       003bb0a89a48ab62d4f57b2c0b6f1a29f099b4a3  0.72  \n",
       "...                                          ...   ...  \n",
       "210868  ffdfa4b3c9a9212774a86a99f39beed20318d1a6  0.02  \n",
       "210869  ffe4c57006027b9016c07f390c505e9bf8c5ab5a  0.90  \n",
       "210870  ffe88b1f2bbe8aaebff80111bde247fefbf090a3  0.02  \n",
       "210871  ffeaf543ce77a973514d395d2df49e94de8ec95e  0.32  \n",
       "210872  fff395754e40f4f71bcd0209715dcca7e6ef7bad  0.32  \n",
       "\n",
       "[210873 rows x 72 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "targets_df,diseases_df,evidence_df = list(load_datasets_with_fallback())\n",
    "display(targets_df)\n",
    "display(diseases_df)\n",
    "display(evidence_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "337371fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "evidence_df = evidence_df[['targetId','diseaseId','score']]\n",
    "c = ['targetId', 'diseaseId']\n",
    "# trgt_disease_with_median = evidence_df.groupby(c)['score'].median()\n",
    "trgt_disease_with_medain_top3   = evidence_df.groupby(c,as_index=False).agg(median = pd.NamedAgg(column='score', aggfunc='median'),\n",
    "                                                      top3   = pd.NamedAgg(column='score', aggfunc= lambda x: (x.nlargest(3).tolist())\n",
    "                                                                              ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "843bc1b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetId</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>median</th>\n",
       "      <th>top3</th>\n",
       "      <th>approvedSymbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>HP_0001249</td>\n",
       "      <td>0.00</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>Orphanet_137</td>\n",
       "      <td>0.32</td>\n",
       "      <td>[0.92, 0.92, 0.9]</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>EFO_0001365</td>\n",
       "      <td>0.32</td>\n",
       "      <td>[0.32, 0.32, 0.32]</td>\n",
       "      <td>CFH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>EFO_1001155</td>\n",
       "      <td>0.02</td>\n",
       "      <td>[0.92, 0.32, 0.32]</td>\n",
       "      <td>CFH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>Orphanet_200421</td>\n",
       "      <td>0.02</td>\n",
       "      <td>[0.02, 0.02, 0.02]</td>\n",
       "      <td>CFH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12057</th>\n",
       "      <td>ENSG00000288702</td>\n",
       "      <td>Orphanet_79235</td>\n",
       "      <td>0.32</td>\n",
       "      <td>[0.32]</td>\n",
       "      <td>UGT1A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12058</th>\n",
       "      <td>ENSG00000288705</td>\n",
       "      <td>EFO_0005556</td>\n",
       "      <td>0.32</td>\n",
       "      <td>[0.9, 0.72, 0.5]</td>\n",
       "      <td>UGT1A5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12059</th>\n",
       "      <td>ENSG00000288705</td>\n",
       "      <td>Orphanet_2312</td>\n",
       "      <td>0.32</td>\n",
       "      <td>[0.9, 0.32, 0.32]</td>\n",
       "      <td>UGT1A5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12060</th>\n",
       "      <td>ENSG00000288705</td>\n",
       "      <td>Orphanet_79234</td>\n",
       "      <td>0.32</td>\n",
       "      <td>[0.9, 0.72, 0.32]</td>\n",
       "      <td>UGT1A5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12061</th>\n",
       "      <td>ENSG00000288705</td>\n",
       "      <td>Orphanet_79235</td>\n",
       "      <td>0.90</td>\n",
       "      <td>[0.9, 0.9, 0.72]</td>\n",
       "      <td>UGT1A5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12062 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              targetId        diseaseId  median                top3  \\\n",
       "0      ENSG00000000419       HP_0001249    0.00               [0.0]   \n",
       "1      ENSG00000000419     Orphanet_137    0.32   [0.92, 0.92, 0.9]   \n",
       "2      ENSG00000000971      EFO_0001365    0.32  [0.32, 0.32, 0.32]   \n",
       "3      ENSG00000000971      EFO_1001155    0.02  [0.92, 0.32, 0.32]   \n",
       "4      ENSG00000000971  Orphanet_200421    0.02  [0.02, 0.02, 0.02]   \n",
       "...                ...              ...     ...                 ...   \n",
       "12057  ENSG00000288702   Orphanet_79235    0.32              [0.32]   \n",
       "12058  ENSG00000288705      EFO_0005556    0.32    [0.9, 0.72, 0.5]   \n",
       "12059  ENSG00000288705    Orphanet_2312    0.32   [0.9, 0.32, 0.32]   \n",
       "12060  ENSG00000288705   Orphanet_79234    0.32   [0.9, 0.72, 0.32]   \n",
       "12061  ENSG00000288705   Orphanet_79235    0.90    [0.9, 0.9, 0.72]   \n",
       "\n",
       "      approvedSymbol  \n",
       "0               DPM1  \n",
       "1               DPM1  \n",
       "2                CFH  \n",
       "3                CFH  \n",
       "4                CFH  \n",
       "...              ...  \n",
       "12057         UGT1A3  \n",
       "12058         UGT1A5  \n",
       "12059         UGT1A5  \n",
       "12060         UGT1A5  \n",
       "12061         UGT1A5  \n",
       "\n",
       "[12062 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined_targets_df = trgt_disease_with_medain_top3.merge(targets_df[['id','approvedSymbol']],left_on='targetId',right_on='id')\n",
    "\n",
    "joined_targets_df.drop(['id'],axis=1,inplace=True)\n",
    "joined_targets_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "04d8082f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetId</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>median</th>\n",
       "      <th>top3</th>\n",
       "      <th>approvedSymbol</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>HP_0001249</td>\n",
       "      <td>0.00</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>Intellectual disability</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004848</td>\n",
       "      <td>HP_0001249</td>\n",
       "      <td>0.00</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>ARX</td>\n",
       "      <td>Intellectual disability</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004961</td>\n",
       "      <td>HP_0001249</td>\n",
       "      <td>0.00</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>HCCS</td>\n",
       "      <td>Intellectual disability</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000005302</td>\n",
       "      <td>HP_0001249</td>\n",
       "      <td>0.72</td>\n",
       "      <td>[0.72]</td>\n",
       "      <td>MSL3</td>\n",
       "      <td>Intellectual disability</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000005339</td>\n",
       "      <td>HP_0001249</td>\n",
       "      <td>0.00</td>\n",
       "      <td>[0.0, 0.0]</td>\n",
       "      <td>CREBBP</td>\n",
       "      <td>Intellectual disability</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12057</th>\n",
       "      <td>ENSG00000284770</td>\n",
       "      <td>Orphanet_2333</td>\n",
       "      <td>0.72</td>\n",
       "      <td>[0.72]</td>\n",
       "      <td>TBCE</td>\n",
       "      <td>Kenny-Caffey syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12058</th>\n",
       "      <td>ENSG00000285053</td>\n",
       "      <td>MONDO_0014968</td>\n",
       "      <td>0.52</td>\n",
       "      <td>[0.72, 0.32]</td>\n",
       "      <td>TBCE</td>\n",
       "      <td>encephalopathy, progressive, with amyotrophy a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12059</th>\n",
       "      <td>ENSG00000285053</td>\n",
       "      <td>Orphanet_93324</td>\n",
       "      <td>0.72</td>\n",
       "      <td>[0.72]</td>\n",
       "      <td>TBCE</td>\n",
       "      <td>Autosomal recessive Kenny-Caffey syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12060</th>\n",
       "      <td>ENSG00000285708</td>\n",
       "      <td>EFO_0009908</td>\n",
       "      <td>0.72</td>\n",
       "      <td>[0.72]</td>\n",
       "      <td>ENSG00000285708</td>\n",
       "      <td>glabellar hemangioma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12061</th>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>MONDO_0012991</td>\n",
       "      <td>0.90</td>\n",
       "      <td>[0.9]</td>\n",
       "      <td>ENSG00000288695</td>\n",
       "      <td>Kahrizi syndrome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12062 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              targetId       diseaseId  median          top3   approvedSymbol  \\\n",
       "0      ENSG00000000419      HP_0001249    0.00         [0.0]             DPM1   \n",
       "1      ENSG00000004848      HP_0001249    0.00         [0.0]              ARX   \n",
       "2      ENSG00000004961      HP_0001249    0.00         [0.0]             HCCS   \n",
       "3      ENSG00000005302      HP_0001249    0.72        [0.72]             MSL3   \n",
       "4      ENSG00000005339      HP_0001249    0.00    [0.0, 0.0]           CREBBP   \n",
       "...                ...             ...     ...           ...              ...   \n",
       "12057  ENSG00000284770   Orphanet_2333    0.72        [0.72]             TBCE   \n",
       "12058  ENSG00000285053   MONDO_0014968    0.52  [0.72, 0.32]             TBCE   \n",
       "12059  ENSG00000285053  Orphanet_93324    0.72        [0.72]             TBCE   \n",
       "12060  ENSG00000285708     EFO_0009908    0.72        [0.72]  ENSG00000285708   \n",
       "12061  ENSG00000288695   MONDO_0012991    0.90         [0.9]  ENSG00000288695   \n",
       "\n",
       "                                                    name  \n",
       "0                                Intellectual disability  \n",
       "1                                Intellectual disability  \n",
       "2                                Intellectual disability  \n",
       "3                                Intellectual disability  \n",
       "4                                Intellectual disability  \n",
       "...                                                  ...  \n",
       "12057                              Kenny-Caffey syndrome  \n",
       "12058  encephalopathy, progressive, with amyotrophy a...  \n",
       "12059          Autosomal recessive Kenny-Caffey syndrome  \n",
       "12060                               glabellar hemangioma  \n",
       "12061                                   Kahrizi syndrome  \n",
       "\n",
       "[12062 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined_target_diesease_df = joined_targets_df.merge(diseases_df[['id','name']],left_on = 'diseaseId',right_on = 'id')\n",
    "joined_target_diesease_df.drop(['id'],axis=1,inplace=True)\n",
    "joined_target_diesease_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c0a8ad7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "joined_target_diesease_df.sort_values(by=['median']).to_json('output.json',orient= 'records')\n",
    "\n",
    "print('Result for part one is now stored as output.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dc759183",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_disease_df = joined_target_diesease_df[['targetId','diseaseId']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aceb0187",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of target-target pairs which share a connection to at least two diseases: 87286\n"
     ]
    }
   ],
   "source": [
    "result = pd.merge(target_disease_df,target_disease_df, on='diseaseId',suffixes =['','_right'])\n",
    "\n",
    "result = result[result.targetId != result.targetId_right]\n",
    "\n",
    "c = ['targetId','targetId_right']\n",
    "\n",
    "target_target_disease_count = result.groupby(c,as_index=False).count()\n",
    "target_target_disease_count = target_target_disease_count[target_target_disease_count.diseaseId >= 2]\n",
    "\n",
    "print('Number of target-target pairs which share a connection to at least two diseases:',len(target_target_disease_count.index))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
